### PR TITLE
GraphExporter: fix replacer methods for nested objects

### PR DESCRIPTION
**Pull request by dunn**
Hopefully unblocks #1071.

From that ticket (https://github.com/samvera/hyrax/pull/1071#issuecomment-305680630):
> Looks like the problem is that subresource_replacer is called when result is falsy (https://github.com/samvera-labs/hyrax/blob/master/app/services/hyrax/graph_exporter.rb#L45), but subresource_replacer itself calls subject_replacer with result as one of the arguments: https://github.com/samvera-labs/hyrax/blob/master/app/services/hyrax/graph_exporter.rb#L59

Since `subject_replacer` only needs `result` in order to compute the class of the object model, just pass that directly as a parameter.

@samvera/hyrax-code-reviewers

### CODE REVIEW

**Comment by dunn** [comment]
Not sure what to make of this failure, since `keyword` is defined in `BasicMetadata` https://travis-ci.org/dunn/hyrax/jobs/266153179#L3046

File: spec/test_app_templates/lib/generators/test_app_generator.rb [hunk 1]
```diff
@@ -13,16 +13,42 @@ def create_generic_work
...
+    generate 'hyrax:work NamespacedWorks::NestedWork'
+    gsub_file 'app/models/namespaced_works/nested_work.rb',
+              'WorkBehavior',
+              <<-EOS.strip_heredoc
+                WorkBehavior
+                  property :created, predicate: ::RDF::Vocab::DC.created, class_name: TimeSpan
+                  accepts_nested_attributes_for :created
+              EOS
+
+    create_file 'app/models/time_span.rb' do
```
Comment (atz) [inline spec/test_app_templates/lib/generators/test_app_generator.rb]: `TimeSpan` should probably be its own template file.

**Comment by dunn** [comment]
Quick summary of what @atz and I discussed on Slack:

His guess was that the error was due to Rails auto-loading `nested_work.rb` prior to the `gsub_file`, leaving it in a half-initialized state.  The proposed solution was to generate it under a different name, make the necessary edits, then move it into place.  But because it's part of a Work generated by Hyrax, that would look something like
```ruby
    generate 'hyrax:work NamespacedWorks::Temp'
    gsub_file 'app/models/namespaced_works/temp.rb',
              'WorkBehavior',
              <<-EOS.strip_heredoc
                WorkBehavior
                  property :created, predicate: ::RDF::Vocab::DC.created, class_name: TimeSpan
                  accepts_nested_attributes_for :created
              EOS

    %w[
      app/actors/hyrax/actors/namespaced_works/temp_actor.rb
      app/controllers/hyrax/namespaced_works/temps_controller.rb
      app/indexers/namespaced_works/temp_indexer.rb
      app/forms/hyrax/namespaced_works/temp_form.rb
      app/presenters/hyrax/namespaced_works/temp_presenter.rb
      app/models/namespaced_works/temp.rb
      app/views/hyrax/namespaced_works/temps/_temp.html.erb
      config/locales/namespaced_works/temp.en.yml
      config/locales/namespaced_works/temp.es.yml
      config/locales/namespaced_works/temp.zh.yml
      config/locales/namespaced_works/temp.de.yml
      config/locales/namespaced_works/temp.fr.yml
      config/locales/namespaced_works/temp.it.yml
      config/locales/namespaced_works/temp.pt-BR.yml
      spec/actors/hyrax/actors/namespaced_works/temp_actor_spec.rb
      spec/controllers/hyrax/namespaced_works/temps_controller_spec.rb
      spec/features/namespaced_works/create_temp_spec.rb
      spec/forms/hyrax/namespaced_works/temp_form_spec.rb
      spec/presenters/hyrax/namespaced_works/temp_presenter_spec.rb
      spec/models/namespaced_works/temp_spec.rb
    ].each do |f|
      gsub_file f, 'Temp', 'NestedWork'
      copy_file f, f.sub('temp', 'nested_work')
      remove_file f
    end

    gsub_file 'config/initializers/hyrax.rb', 'Temp', 'NestedWork'
```
And I haven't gotten that to run without errors yet.  Moreover it's extremely not future-proof.

My next idea was to manually reload the module before the failing test with 
```ruby
  before do
    # https://geminstallthat.wordpress.com/2008/08/11/reloading-classes-in-rspec/
    Object.send(:remove_const, "NamespacedWorks")
    load Rails.root.join("app", "indexers", "namespaced_works", "nested_work_indexer.rb")
    load Rails.root.join("app", "models", "namespaced_works", "nested_work.rb")
  end
```
But that made no difference and we still got the `keyword` error.

Futhermore, just now instead of running `bundle exec rake` to generate the test app and run tests all in one go, I generated the test app with `bundle exec rake engine_cart:regenerate` then started the solr/fcrepo wrappers manually and ran `bundle exec rspec`.  I still got the very same error, suggesting auto-loading isn't the culprit at all:
```ruby
  2) Hyrax::Forms::BatchEditForm#terms
     Failure/Error:
       NamespacedWorks::NestedWork.create!(
         title: ["title 2"],
         keyword: ["123"],
         creator: ["Fred"],
         publisher: ['Rand McNally'],
         language: ['en'],
         resource_type: ['bar'],
         contributor: ['contributor2'],
         description: ['description2'],
         license: ['license2'],

     ActiveTriples::UndefinedPropertyError:
       The property `keyword` is not defined on class 'NamespacedWorks::NestedWork::GeneratedResourceSchema'
     # ./spec/forms/hyrax/forms/batch_edit_form_spec.rb:21:in `block (2 levels) in <top (required)>'
     # ./spec/forms/hyrax/forms/batch_edit_form_spec.rb:38:in `block (2 levels) in <top (required)>'
     # ./spec/forms/hyrax/forms/batch_edit_form_spec.rb:39:in `block (2 levels) in <top (required)>'
     # ./spec/forms/hyrax/forms/batch_edit_form_spec.rb:44:in `block (3 levels) in <top (required)>'
     # ./spec/forms/hyrax/forms/batch_edit_form_spec.rb:47:in `block (3 levels) in <top (required)>'

**Comment by dunn** [comment]
tl;dr i'm stumped

**Comment by atz** [comment]
@dunn Ugh, yeah, that is uglier than I thought and only makes things worse (by generating more ruby files into autoloaded directories that we then attempt to edit).  My hypothetical workaround was to specify a **non-ruby** file destination for the relevant model you want to tweak, but you are right that the Hyrax bundling of generators will make that more difficult.  

The step that `generate hyrax:install` calls is `generate hyrax:models`, with [relevant lines here](https://github.com/samvera/hyrax/blob/master/lib/generators/hyrax/models_generator.rb#L39-L46):
```ruby
  def create_collection
    copy_file 'app/models/collection.rb', 'app/models/collection.rb'
    copy_file 'spec/models/collection_spec.rb', 'spec/models/collection_spec.rb' if rspec_installed?
  end

  def create_file_set
    copy_file 'app/models/file_set.rb', 'app/models/file_set.rb'
    copy_file 'spec/models/file_set_spec.rb', 'spec/models/file_set_spec.rb' if rspec_installed?
  end
```
At this point, maybe just make yourself a template file that has the content you want and call `copy_file`?

**Comment by dunn** [comment]
@no-reply helped me fix those `keyword` errors, but there's a final view spec failure that I don't understand: https://travis-ci.org/samvera/hyrax/jobs/302225032#L3091
```ruby
  1) shared/_select_work_type_modal.html.erb draws the modal

     Failure/Error: data-single="<%= row_presenter.switch_to_new_work_path(route_set: main_app, params: params) %>"

       #<InstanceDouble(Hyrax::SelectTypePresenter) (anonymous)> received unexpected message :switch_to_new_work_path with ({:route_set=>#<ActionDispatch::Routing::RoutesProxy:0x000000001569a630 @scope=#<ActionView::TestCase:...x0000000006a03e48>, @script_namer=nil>, :params=><ActionController::Parameters {} permitted: false>})
     # ./app/views/shared/_select_work_type_modal.html.erb:14:in `block in __home_travis_build_samvera_hyrax_app_views_shared__select_work_type_modal_html_erb___1080083396873507019_179694460'
     # ./app/views/shared/_select_work_type_modal.html.erb:10:in `__home_travis_build_samvera_hyrax_app_views_shared__select_work_type_modal_html_erb___1080083396873507019_179694460'
     # ./spec/views/shared/select_work_type_modal.html.erb_spec.rb:20:in `block (2 levels) in <top (required)>'

**Comment by dunn** [comment]
Oh, it might have just been old code from the rebase.
